// later.js/lib/Document.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _serializeJavascript = _interopRequireDefault(require("serialize-javascript"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * Default Document class that provides some sane defaults.
 *
 * @param {Object}   options               Document options.
 * @param {Object}   options.helmet        Optional React-Helmet override.
 * @param {Object}   options.assets        Asset manifest.
 * @param {Object}   options.data          Preloaded data.
 * @param {String}   options.content       Rendered HTML content.
 * @param {Function} options.appendToHead  Function that returns a React
 *                                         component that will be placed at
 *                                         the end of the document head.
 */
var Document = function Document(_ref) {
  var helmet = _ref.helmet,
      assets = _ref.assets,
      data = _ref.data,
      content = _ref.content,
      appendToHead = _ref.appendToHead;
  var htmlAttrs = helmet.htmlAttributes.toComponent();
  var bodyAttrs = helmet.bodyAttributes.toComponent();
  return _react.default.createElement("html", htmlAttrs, _react.default.createElement("head", null, _react.default.createElement("meta", {
    httpEquiv: "X-UA-Compatible",
    content: "IE=edge"
  }), _react.default.createElement("meta", {
    charSet: "utf-8"
  }), _react.default.createElement("title", null, "later.js \uD83D\uDC4B"), _react.default.createElement("meta", {
    name: "viewport",
    content: "width=device-width, initial-scale=1"
  }), helmet.title.toComponent(), helmet.meta.toComponent(), helmet.link.toComponent(), appendToHead ? appendToHead() : null), _react.default.createElement("body", bodyAttrs, _react.default.createElement("div", {
    id: "root",
    dangerouslySetInnerHTML: {
      __html: content
    }
  }), _react.default.createElement("script", {
    id: "preloaded-state",
    type: "application/json",
    dangerouslySetInnerHTML: {
      __html: (0, _serializeJavascript.default)(_objectSpread({}, data))
    }
  }), _react.default.createElement("script", {
    type: "text/javascript",
    src: assets.client.js,
    defer: true,
    crossOrigin: "anonymous"
  })));
};

var _default = Document;
exports.default = _default;
